PR: Add Panels and PanelsManager, and use it to add LineNumberArea to editor

rlaverde opened the pull request:
Using pyqode code for managing editors panels

Progress
- add pyqode classes
- add panels to codeeditor
- add linenumberarea using panel manager

TODO

- [x] Reconnect all calculation that depends of width linenumberarea (lines commented out)
- [x] Add pyqode copyright and license and links to source files
- [x] Reword, reorder and squash some commits
- [x] Test that everything is working

rlaverde commented:
@goanpeca I think this is kind of ready (It will need an squash and reorder of some commits) could you review it?

And what do you think of the deletion of the` LineNumberPanel.update_width` method? It was causing an infinite recursion, and I think that after adding PanelManager that method in not longer needed

goanpeca commented:
rafa why the tests fail om circleci? do you have the latest master?

ccordoba12 commented:
Tests are failing because CIrcleCI tries to be smart and run without a `circle.yml` file. A simple merge with master will fix that :-)

ccordoba12 commented on spyder/api/manager.py:
```diff
@@ -0,0 +1,53 @@
+        editor.panels.install(MyPanel(), MyPanel.Position.Right)
+        my_panel = editor.panels.get(MyPanel)
+
+        # and so on
```
Please rewrite this docstring to refer to Spyder functionality instead of PyQode one

ccordoba12 commented on spyder/api/manager.py:
```diff
@@ -0,0 +1,53 @@
+    @property
+    def editor(self):
+        """
+        Return a reference to the parent code edit widget.
```
We don't code edit widgets, please rewrite

ccordoba12 commented on spyder/api/manager.py:
```diff
@@ -0,0 +1,53 @@
+
+    def __init__(self, editor):
+        """
+        :param editor: CodeEdit instance to control
```
Please rewrite too

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,132 @@
+        - :meth:`pyqode.core.api.Mode.on_state_changed`
+
+    ..warning: The mode will be identified by its class name, this means that
+    **there cannot be two modes of the same type on the same editor instance!**
```
Please rewrite this docstring too to refer to Spyder objects instead of PyQode ones

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,132 @@
+
+        **READ ONLY**
+
+        :rtype: pyqode.core.api.code_edit.CodeEdit
```
Rewrite

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,132 @@
+        :meth:`pyqode.core.api.Mode.on_state_changed` will be called as soon
+        as the mode state changed.
+
+        :type: bool
```
Rewrite

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,132 @@
+        .. note:: This method is called by editor when you install a Mode.
+                  You should never call it yourself, even in a subclasss.
+
+        .. warning:: Don't forget to call **super** when subclassing
```
Rewrite

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+            method!
+
+        :param editor: editor instance
+        :type editor: pyqode.core.api.CodeEdit
```
Rewrite

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,309 @@
+
+"""
+This module contains the panels controller, responsible of drawing panel
+inside CodeEdit's margins
```
We don't have CodeEdit, please rewrite

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,309 @@
+    """
+    Manages the list of panels and draws them inised the margin of the code
+    edit widget.
+
```
Remove blank

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,309 @@
+    def clear(self):
+        """
+        Removes all panel from the editor.
+
```
Remove blank

ccordoba12 commented:
This looks good to me. I'd only like to see better docstrings for files added in this PR, i.e. removing the references to pyqode objects (like `CodeEdit`) and classes, and use the Spyder ones instead :-)

rlaverde commented:
@ccordoba12 I addressed your comments, but there is a little error, when editor is loaded the panels are drawn over the editor (cropping the first lines), but after scrolling they work fine.

rlaverde commented:
> but there is a little error, when editor is loaded the panels are drawn over the editor (cropping the first lines)

Fixed, It was missing the override of `CodeEditor.showEvent()` method

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,134 @@
+
+    A panel (model child class) is added to an editor by using the
+    PanelsManager:
+        - :meth:`pyqode.core.api.CodeEditor.panels.append`
```
Please change this line to the corresponding behavior in Spyder

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+
+class PanelsManager(Manager):
+    """
+    Manages the list of panels and draws them inised the margin of the
```
Misspelling: inised -> inside

goanpeca commented:
@ccordoba12 ? can we merge this?

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+        #: Specifies whether the panel is dynamic. A dynamic panel is a panel
+        #: that will be shown/hidden depending on the context.
+        #: Dynamic panel should not appear in any GUI menu (e.g. no display
+        #: in the panels menu of the notepad example).
```
Please remove from this comment the reference to *the notepad example* (which we don't have)

ccordoba12 replied:
Also change `#: ` to `#` in all comments

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,134 @@
+    def on_uninstall(self):
+        """
+        Uninstalls the mode from the editor.
+        """
```
Please make this a single line docstring

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,134 @@
+
+    def __init__(self):
+        #: Mode name/identifier. :class:`spyder.widgets.sourcecode.CodeEditor`
+        # uses that as the attribute key when you install a mode.
```
Please make this a docstring

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,134 @@
+    @property
+    def enabled(self):
+        """
+        Tells if the mode is enabled,
```
Finish this line with a dot and add a blank line to separate it from the description

ccordoba12 commented on spyder/api/mode.py:
```diff
@@ -0,0 +1,134 @@
+
+class Mode(object):
+    """
+    Base class for editor extensions. An extension is a "thing" that can be
```
This should look like

    Base class for editor extensions.

    An extension is a "thing" that can be ...

ccordoba12 commented on spyder/api/manager.py:
```diff
@@ -0,0 +1,45 @@
+
+    def __init__(self, editor):
+        """
+        :param editor: CodeEditor instance to control
```
Make this a single class docstring

ccordoba12 commented on spyder/api/manager.py:
```diff
@@ -0,0 +1,45 @@
+    @property
+    def editor(self):
+        """
+        Return a reference to the parent CodeEditor widget.
```
Single line docstring

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+    """
+    class Position(object):
+        """
+        Enumerates the possible panel positions
```
Single line docstring

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+    @property
+    def scrollable(self):
+        """
+        A scrollable panel will follow the editor's scroll-bars. Left and right
```
Change to

    A scrollable panel will follow the editor's scroll-bars.

    Left and right ...

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+            self.palette().windowText().color()))
+
+    def paintEvent(self, event):
+        # Fills the panel background using QPalette
```
Make this a single line docstring

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,127 @@
+
+    def setVisible(self, visible):
+        """
+        Shows/Hides the panel
```
Dot missing at the end

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+class PanelsManager(Manager):
+    """
+    Manages the list of panels and draws them inside the margin of the
+    CodeEditor widget.
```
Change this to

    Manage the list of panels and draw them inside the margins of
    CodeEditor widgets.

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+    def keys(self):
+        """
+        Returns the list of installed panel names.
+        """
```
Single line docstring

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+
+    def values(self):
+        """
+        Returns the list of installed panels.
```
Single line docstring

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+        return list(self._panels[zone].values())
+
+    def refresh(self):
+        """ Refreshes the editor panels (resize and update margins) """
```
Missing dot at the end, remove blanks at the beginning and end of this docstring.

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+                     force_update_margins=True)
+
+    def resize(self):
+        """ Resizes panels """
```
Removes blanks, missing dot

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+            bottom += size_hint.height()
+
+    def _update(self, rect, delta_y, force_update_margins=False):
+        """ Updates panels """
```
Remove blanks, missing dot

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+            self._update_viewport_margins()
+
+    def _update_viewport_margins(self):
+        """ Update viewport margins """
```
Remove blanks, missing dot

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,308 @@
+        return self._margin_sizes[position]
+
+    def _compute_zones_sizes(self):
+        """ Compute panel zone sizes """
```
Remove blanks, missing dot

ccordoba12 commented on spyder/widgets/sourcecode/codeeditor.py:
```diff
@@ -1249,13 +1258,16 @@ def scrollflagarea_paint_event(self, event):
+        self.panels.resize()
+
+    def showEvent(self, event):
+        """ Overrides showEvent to update the viewport margins """
```
Remove blanks, missing dot.

rlaverde replied:
@goanpeca This style issue was in an existing file, and wasn't detected

ccordoba12 commented:
I left a lot of comments that ciocheck didn't catch.

goanpeca commented:
I think it may be a bug on ciocheck I will invetigate further

rlaverde commented:
> I think it may be a bug on ciocheck I will investigate further

Maybe because they are new files

goanpeca commented:
>Maybe because they are new files

Yes thats the thing, I dont think I ever used that as a test case, for commited lines (which is the mode we are using now.) Will add an issue

ccordoba12 commented on spyder/api/panel.py:
```diff
@@ -0,0 +1,125 @@
+
+def _logger():
+    """ Returns module's logger """
+    return logging.getLogger(__name__)
```
Please change this (and all other uses) of this `_logger` function to use our own `debug_print` one.

rlaverde replied:
I haven't notice `_logger` isn't used in the files that I copy, I'm going to delete this functions.

ccordoba12 commented on spyder/widgets/panels/manager.py:
```diff
@@ -0,0 +1,299 @@
+
+    def clear(self):
+        """
+        Removes all panel from the CodeEditor.
```
One line docstring.

ccordoba12 commented:
@rlaverde, I left one last comment, the I'd say this is ready ;-)

goanpeca commented:
Woot :-)
